src/h5web/vis-packs/core/heatmap/utils.ts

```diff
@@ -2,20 +2,34 @@ import { range } from 'lodash-es';
 import type ndarray from 'ndarray';
 import type { D3Interpolator, Dims } from './models';
 
+const GRADIENT_PRECISION = 1 / 20;
+const GRADIENT_RANGE = range(0, 1 + GRADIENT_PRECISION, GRADIENT_PRECISION);
+
 export function getDims(dataArray: ndarray): Dims {
   const [rows, cols] = dataArray.shape;
   return { rows, cols };
 }
 
-export function generateCSSLinearGradient(
+function getColorStops(
   interpolator: D3Interpolator,
-  direction: 'top' | 'bottom' | 'right' | 'left'
+  minMaxOnly: boolean
 ): string {
-  const gradientColors = range(0, 1.1, 0.1)
-    .map(interpolator)
-    .reduce((acc, val) => `${acc},${val}`);
+  if (minMaxOnly) {
+    const min = interpolator(0);
+    const max = interpolator(1);
+    return `${min}, ${min} 50%, ${max} 50%, ${max}`;
+  }
+
+  return GRADIENT_RANGE.map(interpolator).join(', ');
+}
 
-  return `linear-gradient(to ${direction},${gradientColors})`;
+export function getLinearGradient(
+  interpolator: D3Interpolator,
+  direction: 'top' | 'bottom' | 'right' | 'left',
+  minMaxOnly = false
+): string {
+  const colorStops = getColorStops(interpolator, minMaxOnly);
+  return `linear-gradient(to ${direction}, ${colorStops})`;
```

Comment:
Not sure `getColorStops` is necessary. You could have put the early return into `getLinearGradient`:

```js
  if (minMaxOnly) {
    const min = interpolator(0);
    const max = interpolator(1);
    return `linear-gradient(to ${direction}, ${min}, ${min} 50%, ${max} 50%, ${max});
  }

  return `linear-gradient(to ${direction}, GRADIENT_RANGE.map(interpolator).join(', ')`;
```

Feel free to keep it like this though.

Reply:
Yeah, I had this at first, but the returns don't fit nicely on single lines -- I have to add `colorStops` constants above them... And I wasn't a fan of having `linear-gradient(to ${direction},` duplicated 🤷‍♂️